airlock doc (#2082)
microsoft/AzureTRE: docs/azure-tre-features/airlock.md

# Airlock

> In a Trusted Research Environment (TRE) the workspaces represent a security boundary that enables researchers to access data, execute analysis, apply algorithms and collect reports. The airlock capability brings the only mechanisme that allows users to `import` or `export` data and tools in a secure fashion with a manual approval.

Author: Fully Agree. Thanks

> This constitutes the mechanism focused in preventing malware inside TRE and/or TRE workspaces and preventing data exfiltration, while allowing researchers to execute their tasks.

Author: Agree. Rephrased it to highlight the primary focus.

> The airlock feature brings several actions: ingress/egress Mechanism; Data movement; Security gates; Approval mechanism and Notifications. As part of TRE's Safe settings all activity must be tracked for auditiing purpose.

Reviewer: Can we have a flow diagram? I'd move this below the bullets below.

Author: Added a state diagram only. It will be useful next to the process description.
Down below we can add one focusing on the data movement through the storage accounts and respective security boundaries. Makes sense?

> The airlock feature brings several actions: ingress/egress Mechanism; Data movement; Security gates; Approval mechanism and Notifications. As part of TRE's Safe settings all activity must be tracked for auditiing purpose.

Reviewer 2: This sentence is not 100% clear to me.
ingress/egress Mechanism vs Data movement, what the differencce
security gates - malware scanning?

Author: Yes, the idea is to enumerate key aspects from Airlock (and thinking in the future):
- ingress/egress Mechanism: is focused on how you move from an internal security boundary (export) to external location. For export the other way around
- Data Movement is more around how data is moved between several locations for different purpose
- Security Gates was the term we found for any (compute) task that may be brought into the process. Curretnly we have malware scanning only. In the future you could have PII scanning and others.

Let me rephrase it to make it clearer.

The Airlock feature aims to address these goals:

> * Prevent unauthorised data export (or import).

Reviewer: ```suggestion
* Prevent unauthorised data import or export.
```

> * Provide a process to allow approved data into the security boundary of a TRE Workspace.

Reviewer: ```suggestion
* Provide a process to allow approved data to be imported through the security boundary of a TRE Workspace.
```

> * TRE provides functionality to track requests and decisions, supporting cycle of revision, approval or rejection.

Author: that's the term used in the definition of airlock feature. It's all about providing way to import/export data while going through a cycle of revision by the airlock managers.

> * Data being imported with an airlock import process must be automatically scanned for security purpose.

Reviewer: ```suggestion
* Data being imported with an airlock import process can be automatically scanned for malware.
```

> * Data bein exported or imported must ne manually reviewed bu the Workspace Owner.

Reviewer 2: ```suggestion
* Data being exported or imported must be manually reviewed by the Workspace Owner.
```

> * Data bein exported or imported must ne manually reviewed bu the Workspace Owner.

Author: The role will be defined in near future. Documentation wise, it makes sense to mention the Airlock Manager name.

* Notify the involved users of the process progress and/or required actions

> * All airlock processes are audited and tracked

Reviewer: ```suggestion
* All steps within the airlock process are audited.
```

> Typically in a TRE, the Airlock feature would be used to allow a researcher to export patient identifiers for an exact cohort without the underlying data ever being exposed. Through the airlock, data to be exported is restricted by a manual review ensuring that no data is exfiltered from TRE.

Reviewer: Not sure we need this line the statement isn't really correct - usually is for exporting outputs of a research project, such as summary results.

Author: Goal was to provide an example for the use of Airlock. Rephrased it, but we can remove it.

> Typically in a TRE, the Airlock feature would be used to allow a researcher to export patient identifiers for an exact cohort without the underlying data ever being exposed. Through the airlock, data to be exported is restricted by a manual review ensuring that no data is exfiltered from TRE.

Reviewer 2: I think this sentence is redundant. seems like a private case. and regarding the review, it was already said before

Author: This sentence highlights a common business scenario for TREs where Airlock is a key feature enabling that scenario.
The goal was to highlight the **human** reviewal. Rephrased it. Thanks!

> The Airlock feature will create events on every miningful step of the processes. This will enable increased flexibility by allowing an organization to extend several underlying mechanisms.

Reviewer: ```suggestion
The Airlock feature will create events on every meaningful step of the processes. This will enable increased flexibility by allowing an organization to extend several underlying mechanisms.
```

> The Airlock feature will create events on every miningful step of the processes. This will enable increased flexibility by allowing an organization to extend several underlying mechanisms.

Reviewer 2: ```suggestion
The Airlock feature will create events on every meaningful step of the processes. This will enable increased flexibility by allowing an organization to extend the notification mechanism.
```

## Ingress/Egress Mechanism

> The Ingress/Egress Mechanism allows a TRE user to start `import` or `export` process to a given workspace. The mechanism provides a set of milestones that must be reached in order to execute a process end to end. This milestones end up to be states:

Reviewer: ```suggestion
The Airlock allows a TRE user to start `import` or `export` process to a given workspace. A number of milestones must be reached in order to complete a successful import or export. These milestones are defined using the following states:
```

Author: Yep. Like it better. Thanks

a. **Draft**: An import/export process has been created but has not yet started. The TRE User/Researcher has now access to a storage location and he must identify the data to be processed. At this point the airlock import/export processes allow a single file to be processed. However a compressed file may be used (zip).
b. **Submitted**: The request was submitted by the ressearcher (not yet processed).

> c. **In-Review**: The request was submitted is in review.

Reviewer 2: ```suggestion
c. **In-Review**: The request is ready to be reviewed. This state can be reached directly from Submitted state or after being scanned for malware (and found clean) 
```

Author: THanks. Just rephrased it to remove malware scanning (for now). Using the term security scan

> d. **Approval In-progress**: The import/export process has been approved, however data movement is still ongoing.

Reviewer: ```suggestion
d. **Approval In-progress**: The Airlock request has been approved, however data movement is still ongoing.
```

Reviewer: Can we change import/export process to Airlock request for the below too?

e. **Approved**: The import/export process has been approved. At this state, data has been securely verified and manually reviewed. The data is now in its final location. For an import process the data is now available in the TRE workspace. It can be accessed by the requestor from within the workspace.
f. **Rejection In-progress**: The import/export process has been rejected, however data movement is still ongoing.
g. **Rejected**: The import/export process has been rejected. The data in the process was rejected manually by the Airlock Manager.
h. **Cancelled**: The import/export process was manually cancelled by the requestor TRE user, a Workspace owner or a TRE administrator. The cancelation is only allowed when the request is not actively changing (i.e. **Draft** or **In-Review** state).
i. **Blocking In-progress**: The import/export process has been blocked, however data movement is still ongoing.

> j. **Blocked By Scan**: The import/export process has been blocked. The security analysis found issues in the submitted data, and consequently quarantined the data.

Reviewer 2: looks good

> When an airlock process is created the state is **Draft** and thre request infrastructure will get created providing a single container to centralize data in the request. Once the processThe user will get a link for this container inside the storage account (URL + SAS token) that he can use to upload the desired data to be processed (import or export).

Author: fixed.Thanks

> When an airlock process is created the state is **Draft** and the request infrastructure will get created providing a single container to centralize data in the request. Once completed, the user user will be able to get a link for this container inside the storage account (URL + SAS token) that he can use to upload the desired data to be processed (import or export).

Reviewer: ```suggestion
When an airlock process is created the initial state is **Draft** and the required infrastructure will get created providing a single container to isolate the data in the request. Once the required infrastructure is provisioned, the user will be able to get a link for this container inside the storage account (URL + SAS token) they can upload the desired data to be processed (import or export).
```

This storage location is external for import (`stalimex`) or internal for export (`stalexint`), however only accessible to the requestor (ex: a TRE user/researcher).
The user will be able to upload a file to the provided storage location, using any tool of their preference: [Azure Storage Explorer](https://azure.microsoft.com/en-us/features/storage-explorer/) or [AzCopy](https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azcopy-v10) which is a command line.

The user Submits the request (TRE API call) starting the data movement (to the `stalimip` - import in-progress or `stalexip` - export in-progress). The airlock request is now in state **Submitted**.

> If enabled, the Security Scanning is started. In case security flaws are found, the request state becomes **Blocking In-progress** while the data is moved to storage rejected (either rejected import `stalimrej` or rejected export `stalexrej`). In this case, the request is finalized with the state **Blocked By Scan**.

Reviewer 2: ```suggestion
  If enabled, the Security Scanning is started. In case security flaws are found, the request state becomes **Blocking In-progress** while the data is moved to blocked storage  (either import blocked `stalimblocked` or export blocked`stalexblocked`). In this case, the request is finalized with the state **Blocked By Scan**.
```

> If the Security Scanning does not identify security flaws, the request state becomes **In-Review**. Simultaneously a  notification is sent to the Airlock Manager user providing the access to data (SAS token + URL with READ permission).

Reviewer 2: ```suggestion
If the Security Scanning does not identify security flaws, the request state becomes **In-Review**. Simultaneously a notification is sent to the Airlock Manager user. The user need to ask for the container url using the TRE API (SAS token + URL with READ permission).
```

> The Security Scanning can be disabled, chaging the request state from **Submitted** straight to **In-Review**.

The Airlock Manager will manually review the data using the tools of their choice available in the TRE workspace. Once review is completed, the Airlock Manager will have to *Approve* or *Reject* the airlock proces, though a TRE API call.

> At this point, the request will change state to either **Approval In-progress** or **Rejection In-progress**, while the data movement occurs moving afterwards to **Approved** or **Rejected** accordingly. The data will now be in the final storage destination: `stalexapp` - external approved export or `stalimapp` - internal approved import.

Reviewer 2: ```suggestion
At this point, the request will change state to either **Approval In-progress** or **Rejection In-progress**, while the data movement occurs moving afterwards to **Approved** or **Rejected** accordingly. The data will now be in the final storage destination: `stalexapp` - export approved  or `stalimapp` - import approved.
```

In an export process data will transition from internal locations (available to the requestor) to public locations in the TRE, after going through a manual review.

Considering that the Airlock requests may require large data movements, the operations can have longer durations, hence becoming the operations asynchronous. This is why states like **Approval In-progress**, **Rejection In-progress** or **Blocking In-progress** will be set while there are data movement operations.

> The data movement mechanism is data-driven, allowing an organization to extend how request data transitions between

## Security Scan

The identified data in a airlock proces, will be submited to a security scan. If the security scan identifies issues the data is quarantined, and a report is added to the process metadata. Both the requestor and Workspace Owner are notified. For successful security scan, the data will remain in state **In-progress**, and accessible to the Workspace Owner.

> * The Security scan will be optional, behind a feature flag, enabled by script

> > * The outcome of security scan will be either the in-progress (`stalexip`) storage or rejected (`stalexrej`)

Reviewer 2: ```suggestion
> * The outcome of security scan will be either the in-progress (`stalexip`) storage or blocked (`stalexblocked`)
```

> * An airlock process will guarantee that the content being imported/exported is secure. It is envisioned that a set of **security gates** are identified to be executed successfully for a process to be approve.

## Approval mechanism

The approval mechanism, is bundled with any airlock process, providing a specific way to `approve` or `reject` the data. This mechanism will allow the Airlock Managers to explicitly approve/reject the process, after having acess to the data. The Airlock Manager users will be able to execute a manual review on the data using the tools available to him in a reviewal TRE Workspace.
Once this manual review is executed, Airlock Managers can proactivelly approve or reject the airlock request.

> The only responsability of the Approval mechanism is to support a cycle of revision, approval or rejection, while tracking the decision.

Author: rephrased it. It's all about the goal.

This mechanism will provice access to the data in the airlock process, and will be able to use a VM in TRE workspace. The data review will be the Airlock Manager responsability

> * It is envisioned that this mechanism to be more flexible and extensible.
> * The `Airlock Manager` is a role defined at workspace instance level and assigned to identities. Initially the `Owner` role will be used.

## Notifications

Throughout the airlock process, the notification mechanism will notify the relevant people to the process. Both the requestor (TRE User/Researcher) and the Workspace Owner will be notified by email, of the relevant process events.

> Whenever the airlock process changes to a state of **Draft**, **Submitted**, **Approved**, **Rejected** , **Blocked By Scan** or **Cancelled**, the process requestor gets notified.

Reviewer 2: missing the x-inprogress states...

When the state changes to `In-progress` the Workspace Onwer (Airlock Manager) gets notified.

> > * The Notification mechanism is also data-driven, allowing an organization to extend the notifications behavior. The mechanism is exemplified with a Logic App determining the notifications logic.

Author: Agree

* `stalimip` - storage (st) airlock (al) import (im) in-progress (ip)
* `stalimrej` - storage (st) airlock (al) import (im) rejected (rej)
* `stalimblocked` - storage (st) airlock (al) import (ex) approved (app)
* `stalexapp` - storage (st) airlock (al) export (ex) approved (app)

Workspace
* `stalimapp` - workspace storage (st) airlock (al) import (im) approved (app)
* `stalexint` - workspace storage (st) airlock (al) export (ex) internal (int)
* `stalexip` - workspace storage (st) airlock (al) export (ex) in-progress (ip)
* `stalexrej` - workspace storage (st) airlock (al) export (ex) rejected (rej)
* `stalexblocked` - workspace storage (st) airlock (al) export (ex) rejected (rej)

> * The external storage accounts (`stalimex`, `stalexapp`), are not bound to any vnet and accessible (with SAS token) via internet.
> * The internal storage account (`stalexint`) is bound to the workspace vnet, so ONLY TRE Users/Researchers on that workspace can access it
> * The (export) in-progress storage account (`stalexip`) is bound to the workspace vnet
> * The (export) blocked storage account (`stalexblocked`) is bound to the workspace vnet

> > * The (export) rejected storage account (`stalexrej`) is bound to the TRE CORE vnet

Reviewer 2: ```suggestion
> * The (export) rejected storage account (`stalexrej`) is bound to the workspace vnet
```

# Airlock

> In a Trusted Research Environment (TRE) the workspaces represent a security boundary that enables researchers to access data, execute analysis, apply algorithms and collect reports. The airlock capability brings the only mechanisme that allows users to `import` or `export` data, tools or other file based artefacts in a secure fashion with a human approval.

Reviewer: ```suggestion
In a Trusted Research Environment (TRE) the workspaces represent a security boundary that enables researchers to access data, execute analysis, apply algorithms and collect reports. The airlock capability brings the only mechanism that allows users to `import` or `export` data, tools or other file based artefacts in a secure fashion with a human approval.
```
